src/query/atomWithQuery.ts

```diff
@@ -28,76 +28,60 @@ export function atomWithQuery<
       ) => AtomQueryOptions<TQueryFnData, TError, TData, TQueryData>),
   equalityFn: (a: TData, b: TData) => boolean = Object.is
 ): WritableAtom<TData, Action> {
-  const stateAtom = atom(() => {
-    const state = {
-      resolve: null as ((data: TData) => void) | null,
-      setData: null as ((data: TData) => void) | null,
-      prevData: null as TData | null,
-      unsubscribe: null as (() => void) | null,
-      handle(result: QueryObserverResult<TData, TError>) {
+  const queryDataAtom = atom(
+    (get) => {
+      const queryClient = get(queryClientAtom)
+      const options =
+        typeof createQuery === 'function' ? createQuery(get) : createQuery
+      let resolve: ((data: TData) => void) | null = null
+      const dataAtom = atom<TData | Promise<TData>>(
+        new Promise<TData>((r) => {
+          resolve = r
+        })
+      )
+      let setData: (data: TData) => void = () => {
+        throw new Error('setting data without mount')
+      }
+      let prevData: TData | null = null
+      const listener = (result: QueryObserverResult<TData, TError>) => {
         // TODO error handling
         if (
           result.data === undefined ||
-          (this.prevData !== null && equalityFn(this.prevData, result.data))
+          (prevData !== null && equalityFn(prevData, result.data))
         ) {
           return
         }
-        this.prevData = result.data
-        if (this.resolve) {
-          this.resolve(result.data)
-          this.resolve = null
-        } else if (this.setData) {
-          this.setData(result.data)
+        prevData = result.data
+        if (resolve) {
+          resolve(result.data)
+          resolve = null
         } else {
-          throw new Error('setting data without mount')
+          setData(result.data)
         }
-      },
-    }
-    return state
-  })
-  const initAtom = atom(
-    (get) => {
-      const options =
-        typeof createQuery === 'function' ? createQuery(get) : createQuery
-      const state = get(stateAtom)
-      const dataAtom = atom<TData | Promise<TData>>(
-        new Promise<TData>((resolve) => {
-          state.resolve = resolve
+      }
+      const defaultedOptions = queryClient.defaultQueryObserverOptions(options)
+      if (typeof defaultedOptions.staleTime !== 'number') {
+        defaultedOptions.staleTime = 1000
```

Comment:
why is this important?

Reply:
Without this, it will invoke fetchFn again on mount. `1000` is just a rough estimate, I guess. I just took the value from `useQuery`.